libraries/chain/include/eosio/chain/controller.hpp

```diff
@@ -326,24 +326,12 @@ namespace eosio { namespace chain {
 
          static std::optional<uint64_t> convert_exception_to_error_code( const fc::exception& e );
 
-         signal<void(uint32_t)>                        block_start; // block_num
-         signal<void(const signed_block_ptr&)>         pre_accepted_block;
-         signal<void(const block_state_legacy_ptr&)>   accepted_block_header;
-         signal<void(const block_state_legacy_ptr&)>   accepted_block;
-         signal<void(const block_state_legacy_ptr&)>   irreversible_block;
+         signal<void(uint32_t)>                        block_start;
+         signal<void(std::tuple<const signed_block_ptr&, const block_id_type&, const signed_block_header&, uint32_t>)> accepted_block_header;
+         signal<void(std::tuple<const signed_block_ptr&, const block_id_type&, const signed_block_header&, uint32_t>)> accepted_block;
+         signal<void(std::tuple<const signed_block_ptr&, const block_id_type&, uint32_t>)> irreversible_block;
```

Comment:
I would create a type:
`using block_signal_params_t = std::tuple<const signed_block_ptr&, const block_id_type&, const signed_block_header&, uint32_t>;`

And use the same type for all 3 signals.

Reply:
Thanks. Will change.

Reply:
Why `signed_block_header`. When is `signed_block_header != static_cast<signed_block_header>(*signed_block_ptr)` ?

Reply:
I see no reason for `uint32_t` block_num. The user can just use `signed_block_ptr->block_num()`. The id is included because `calculate_id()` is expensive, but `block_num()` I think is efficient enough we don't need to include it in the signal.

Reply:
I decided to use `signed_block_header` and `block_num` explicitly for calling out parameters required by signal clients and for easier review to make sure nothing is missed during the changes. But we need to strike some balances between the number of parameters to signal and abstraction leakage.

Reply:
https://github.com/AntelopeIO/leap/pull/1996/commits/7c4ca8a5df4d7b5438c03663ddbbb1ad9951b8d0